web/fido_app/models.py

```diff
@@ -31,14 +31,25 @@ class User(db.Model, UserMixin):
     display_name = db.Column(db.String(80))
     high_score = db.Column(db.Integer, default=0)
 
+    # Saves the last date the user logged in with both a password and FIDO2
+    last_complete_login = db.Column(db.Date)
+    # Each bit, starting with the least significant bit,
+    # represents a new day since account registration.
+    # A 1 represents a day in which the user logged in;
+    # a 0 represents a skipped day.
+    # E.g., if a user registers on Mar. 1st and then
+    # logs in on the 3rd, 4th, and 6th, `login_bitfield`
+    # should store 22 (10110 in binary).
+    login_bitfield = db.Column(db.Integer, default=0)
```

Comment:
I did consider this, but I decided it was fine when I initially implemented it because, like you said, it won't affect our study. Normal participants would only use up 8 bits, and if someone logged in every day that they could for some reason, that would still be limited to how long we keep the service online, which I imagine would be like 2 weeks. I know this code wouldn't generalize to experiments longer than a month, but a BigInteger seemed like a waste of space. If that's important, I can change it.

Reply:
Yeah, maybe we just want to add a comment here explaining that compensation tracking will fail for studies longer than 32 days.

Reply:
Fair enough. Added in [1470b03](https://github.com/team-pass/FIDO-login/pull/70/commits/1470b03606a2f83dfb4142120eb9fab29482c09d).